hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/LazySet.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.replicatedmap.impl.record;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+class LazySet<K, V, R>
+        implements Set<R> {
+
+    private final InternalReplicatedMapStorage<K, V> storage;
+    private final IteratorFactory<K, V, R> iteratorFactory;
+
+    public LazySet(IteratorFactory<K, V, R> iteratorFactory, InternalReplicatedMapStorage<K, V> storage) {
+
+        this.iteratorFactory = iteratorFactory;
+        this.storage = storage;
+    }
+
+    @Override
+    public int size() {
+        return storage.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return storage.isEmpty();
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        throw new UnsupportedOperationException("LazySet does not support contains requests");
```

Comment:
I looked into it and I don't think we should implement it, it is O(n) if you call this method, if a user calls it more often it will be horrible for the applications runtime. It is better to throw the exception and let the user copy the result set.